src/compiler/binder.ts

```diff
@@ -1655,20 +1658,41 @@ namespace ts {
         }
 
         function bindBindingElementFlow(node: BindingElement) {
-            if (isBindingPattern(node.name)) {
-                // When evaluating a binding pattern, the initializer is evaluated before the binding pattern, per:
-                // - https://tc39.es/ecma262/#sec-destructuring-binding-patterns-runtime-semantics-iteratorbindinginitialization
-                //   - `BindingElement: BindingPattern Initializer?`
-                // - https://tc39.es/ecma262/#sec-runtime-semantics-keyedbindinginitialization
-                //   - `BindingElement: BindingPattern Initializer?`
-                bind(node.dotDotDotToken);
-                bind(node.propertyName);
+            // When evaluating a binding pattern, the initializer is evaluated before the binding pattern, per:
+            // - https://tc39.es/ecma262/#sec-destructuring-binding-patterns-runtime-semantics-iteratorbindinginitialization
+            //   - `BindingElement: BindingPattern Initializer?`
+            // - https://tc39.es/ecma262/#sec-runtime-semantics-keyedbindinginitialization
+            //   - `BindingElement: BindingPattern Initializer?`
+            bind(node.dotDotDotToken);
+            bind(node.propertyName);
+            bindOptionalFlow(() => {
                 bind(node.initializer);
-                bind(node.name);
-            }
-            else {
-                bindEachChild(node);
+            });
+            bind(node.name);
+        }
+
+        function bindParameterFlow(node: ParameterDeclaration) {
+            node.modifiers?.forEach(bind);
+            bind(node.dotDotDotToken);
+            bind(node.questionToken);
+            bind(node.type);
+            bindOptionalFlow(() => {
+                bind(node.initializer);
+            });
+            bind(node.name);
+        }
+
+        // a BindingElement/Parameter does not have side effects if initializers are not evaluated and used. (see GH#49759)
+        function bindOptionalFlow(cb: () => void) {

```

Comment:
Rather than this callback situation, can you just declare `bindInitializer`, inline the bind call where `cb()` is, then call `bindInitializer(node.initializer)` instead? Given this helper is only used in this one situation (as the comment implies), I'm not sure it's worth the abstraction quite yet.